fuse_loss/include/fuse_loss/qwt_loss_plot.h

```diff
@@ -0,0 +1,333 @@
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef FUSE_LOSS_QWT_LOSS_PLOT_H
+#define FUSE_LOSS_QWT_LOSS_PLOT_H
+
+
+#include <qwt_legend.h>
+#include <qwt_plot.h>
+#include <qwt_plot_curve.h>
+#include <qwt_plot_grid.h>
+#include <qwt_plot_magnifier.h>
+#include <qwt_plot_panner.h>
+#include <qwt_plot_renderer.h>
+#include <qwt_plot_zoomer.h>
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+
+namespace fuse_loss
+{
+
+/**
+ * @brief HSV colormap that varies the hue H in the [0, 1) range with equidistant points for the given size.
+ * The hue H interval is open because the hue is the same for 0 and 1. Therefore, we never take hue == 1.
+ */
+class HSVColormap
+{
+public:
+  explicit HSVColormap(const size_t size = 256)
+  {
+    if (size == 0)
+    {
+      return;
+    }
+
+    colormap_.reserve(size);
+
+    double hue = 0.0;
+    const double hue_increment = 1.0 / size;
+    for (size_t i = 0; i < size; ++i, hue += hue_increment)
+    {
+      QColor color;
+      color.setHsvF(hue, 1.0, 1.0);
+
+      colormap_.push_back(color);
+    }
+  }
+
+  const QColor& operator[](const size_t i) const
+  {
+    return colormap_[i];
+  }
+
+  size_t size() const
+  {
+    return colormap_.size();
+  }
+
+private:
+  std::vector<QColor> colormap_;
+};
+
+class LossEvaluator
+{
+public:
+  explicit LossEvaluator(const std::vector<double>& residuals) : residuals_(residuals)
+  {
+  }
+
+  std::vector<double> rho(const ceres::LossFunction* loss_function) const
```

Comment:
I was lazy here and didn't re-use the code of these methods. In principle I should be able to only pass the lambda that is used in `std::transform`. I'm just not sure what would be the argument type for that though.